Class annotated with Introspected.Builder as Lombok builder failing to compile

### Expected Behavior

Class
```
package com.example;

import io.micronaut.core.annotation.Introspected;
import io.micronaut.serde.annotation.Serdeable;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder(builderClassName = "Builder")
@Serdeable
@NoArgsConstructor
@AllArgsConstructor
@Introspected(builder = @Introspected.IntrospectionBuilder(builderClass = RobotCpuData.Builder.class))
public final class RobotCpuData {
    private String id;

    private String manufacturer;
}
```
 can be compiled

### Actual Behaviour

Compilation fails with the error
```
RobotCpuData.java:16: error: Builder class not found on compilation classpath: <error>
public final class RobotCpuData {
             ^
```

### Steps To Reproduce

Just add class above to the project and try to compile.

### Environment Information

Any

### Example Application

_No response_

### Version

4.6.x

--- a/core-processor/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadataBuilder.java
+++ b/core-processor/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadataBuilder.java
@@ -1568,6 +1568,14 @@ public abstract class AbstractAnnotationMetadataBuilder<T, A> {
     }
 
     /**
+     * Used to clear mutated metadata at the end of a compilation cycle.
+     */
+    @Internal
+    public static void clearMutated(@NonNull Object key) {
+        MUTATED_ANNOTATION_METADATA.remove(key);
+    }
+
+    /**
      * Used to clear caches at the end of a compilation cycle.
      */
     @Internal

--- a/core/src/main/java/io/micronaut/core/annotation/AnnotationClassValue.java
+++ b/core/src/main/java/io/micronaut/core/annotation/AnnotationClassValue.java
@@ -85,7 +85,7 @@ public final class AnnotationClassValue<T> implements CharSequence, Named {
     /**
      * Constructs a class value for a type that is present.
      *
-     * @param instance The instnace
+     * @param instance The instance
      * @since 1.1
      */
     @SuppressWarnings("unchecked")
@@ -107,6 +107,9 @@ public final class AnnotationClassValue<T> implements CharSequence, Named {
      * @since 1.1
      */
     public @NonNull Optional<T> getInstance() {
+        if (instance instanceof AnnotationClassValue.UnresolvedClass unresolvedClass) {
+            throw unresolvedClass.error;
+        }
         return Optional.ofNullable(instance);
     }
 
@@ -117,7 +120,18 @@ public final class AnnotationClassValue<T> implements CharSequence, Named {
      * @since 1.1
      */
     public boolean isInstantiated() {
-        return instantiated || getInstance().isPresent();
+        return instantiated ||
+            (getInstance().isPresent() && !(getInstance().get() instanceof AnnotationClassValue.UnresolvedClass));
+    }
+
+    /**
+     * fail with an error if unresolved.
+     */
+    @Internal
+    void failIfError() {
+        if (instance instanceof AnnotationClassValue.UnresolvedClass unresolvedClass) {
+            throw unresolvedClass.error;
+        }
     }
 
     /**
@@ -170,4 +184,11 @@ public final class AnnotationClassValue<T> implements CharSequence, Named {
     public int hashCode() {
         return name.hashCode();
     }
+
+    /**
+     * Represented an unresolved type.
+     * @param error The error that should be thrown
+     */
+    @Internal
+    public record UnresolvedClass(RuntimeException error) { }
 }

--- a/core/src/main/java/io/micronaut/core/annotation/AnnotationValue.java
+++ b/core/src/main/java/io/micronaut/core/annotation/AnnotationValue.java
@@ -723,11 +723,14 @@ public class AnnotationValue<A extends Annotation> implements AnnotationValueRes
         }
         Object o = values.get(member);
         if (o instanceof AnnotationClassValue<?> annotationClassValue) {
+            annotationClassValue.failIfError();
             return Optional.of(annotationClassValue);
         }
         if (o instanceof AnnotationClassValue<?>[] annotationClassValues) {
             if (annotationClassValues.length > 0) {
-                return Optional.of(annotationClassValues[0]);
+                AnnotationClassValue<?> acv = annotationClassValues[0];
+                acv.failIfError();
+                return Optional.of(acv);
             }
         }
         if (o instanceof String className) {

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -34,6 +34,7 @@ jcache = "1.1.1"
 junit5 = "5.10.3"
 junit-platform="1.10.3"
 logback = "1.5.12"
+lombok = "1.18.34"
 logbook-netty = "3.7.0"
 log4j = "2.23.1"
 micronaut-aws = "4.7.1"
@@ -232,6 +233,7 @@ log4j = { module = "org.apache.logging.log4j:log4j-core", version.ref = "log4j" 
 logback-classic = { module = "ch.qos.logback:logback-classic", version.ref = "logback" }
 
 logbook-netty = { module = "org.zalando:logbook-netty", version.ref = "logbook-netty" }
+lombok = { module = "org.projectlombok:lombok", version.ref = "lombok" }
 
 micronaut-docs = { module = "io.micronaut.docs:micronaut-docs-asciidoc-config-props", version.ref = "micronaut-docs" }
 micronaut-build-plugins = { module = "io.micronaut.build.internal:micronaut-gradle-plugins", version.ref="micronaut-build-plugins"}

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/AbstractInjectAnnotationProcessor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/AbstractInjectAnnotationProcessor.java
@@ -23,6 +23,7 @@ import io.micronaut.core.util.CollectionUtils;
 import io.micronaut.inject.annotation.AbstractAnnotationMetadataBuilder;
 import io.micronaut.inject.visitor.TypeElementVisitor;
 
+import java.util.LinkedHashMap;
 import javax.annotation.processing.AbstractProcessor;
 import javax.annotation.processing.Filer;
 import javax.annotation.processing.Messager;
@@ -74,7 +75,7 @@ abstract class AbstractInjectAnnotationProcessor extends AbstractProcessor {
     protected MutableConvertibleValues<Object> visitorAttributes = new MutableConvertibleValuesMap<>();
     protected AnnotationProcessingOutputVisitor classWriterOutputVisitor;
     protected JavaVisitorContext javaVisitorContext;
-    protected Set<String> postponedTypes = new LinkedHashSet<>();
+    protected Map<String, Object> postponedTypes = new LinkedHashMap<>();
     private boolean incremental = false;
     private final Set<String> supportedAnnotationTypes = new HashSet<>(5);
     private final Map<String, Boolean> isProcessedCache = new HashMap<>(30);
@@ -219,7 +220,7 @@ abstract class AbstractInjectAnnotationProcessor extends AbstractProcessor {
             filer,
             visitorAttributes,
             getVisitorKind(),
-            postponedTypes
+            postponedTypes.keySet()
         );
     }
 

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/JavaAnnotationMetadataBuilder.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/JavaAnnotationMetadataBuilder.java
@@ -371,12 +371,24 @@ public class JavaAnnotationMetadataBuilder extends AbstractAnnotationMetadataBui
             final MetadataAnnotationValueVisitor resolver = new MetadataAnnotationValueVisitor(originatingElement, (ExecutableElement) member, resolvedDefaults);
             value.accept(resolver, this);
             Object resolvedValue = resolver.resolvedValue;
+
             if (resolvedValue != null) {
-                if (isEvaluatedExpression(resolvedValue)) {
-                    resolvedValue = buildEvaluatedExpressionReference(originatingElement, annotationName, memberName, resolvedValue);
+                if ("<error>".equals(resolvedValue) &&
+                    Class.class.getName().equals(this.modelUtils.resolveTypeName(((ExecutableElement) member).getReturnType()))) {
+                    resolvedValue = new AnnotationClassValue<>(
+                        new AnnotationClassValue.UnresolvedClass(new PostponeToNextRoundException(
+                            originatingElement,
+                            originatingElement.getSimpleName().toString() + "@" + annotationName + "(" + memberName + ")"
+                        ))
+                    );
+                    annotationValues.put(memberName, resolvedValue);
+                } else {
+                    if (isEvaluatedExpression(resolvedValue)) {
+                        resolvedValue = buildEvaluatedExpressionReference(originatingElement, annotationName, memberName, resolvedValue);
+                    }
+                    validateAnnotationValue(originatingElement, annotationName, member, memberName, resolvedValue);
+                    annotationValues.put(memberName, resolvedValue);
                 }
-                validateAnnotationValue(originatingElement, annotationName, member, memberName, resolvedValue);
-                annotationValues.put(memberName, resolvedValue);
             }
         }
     }

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/TypeElementVisitorProcessor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/TypeElementVisitorProcessor.java
@@ -27,6 +27,7 @@ import io.micronaut.core.order.OrderUtil;
 import io.micronaut.core.util.CollectionUtils;
 import io.micronaut.core.util.StringUtils;
 import io.micronaut.core.version.VersionUtils;
+import io.micronaut.inject.annotation.AbstractAnnotationMetadataBuilder;
 import io.micronaut.inject.ast.ConstructorElement;
 import io.micronaut.inject.ast.ElementQuery;
 import io.micronaut.inject.ast.EnumConstantElement;
@@ -240,7 +241,10 @@ public class TypeElementVisitorProcessor extends AbstractInjectAnnotationProcess
                 roundEnv.getRootElements()
             ).filter(notGroovyObject).forEach(elements::add);
 
-            postponedTypes.stream().map(elementUtils::getTypeElement).filter(Objects::nonNull).forEach(elements::add);
+            for (Object nativeType : postponedTypes.values()) {
+                AbstractAnnotationMetadataBuilder.clearMutated(nativeType);
+            }
+            postponedTypes.keySet().stream().map(elementUtils::getTypeElement).filter(Objects::nonNull).forEach(elements::add);
             postponedTypes.clear();
 
             if (!elements.isEmpty()) {
@@ -271,7 +275,7 @@ public class TypeElementVisitorProcessor extends AbstractInjectAnnotationProcess
                             }
                             error(originatingElement.element(), e.getMessage());
                         } catch (PostponeToNextRoundException e) {
-                            postponedTypes.add(javaClassElement.getName());
+                            postponedTypes.put(javaClassElement.getName(), e.getErrorElement());
                         }
                     }
                 }

--- a/test-suite/build.gradle
+++ b/test-suite/build.gradle
@@ -63,7 +63,8 @@ dependencies {
     testImplementation libs.testcontainers.spock
     // end::testcontainers-dependencies[]
     testImplementation libs.awaitility
-
+    testImplementation(libs.lombok)
+    testAnnotationProcessor(libs.lombok)
     testAnnotationProcessor project(":test-suite-helper")
     testAnnotationProcessor project(":inject-java")
     testAnnotationProcessor project(":test-suite")

--- a/test-suite/src/test/java/io/micronaut/test/lombok/LombokIntrospectedBuilderTest.java
+++ b/test-suite/src/test/java/io/micronaut/test/lombok/LombokIntrospectedBuilderTest.java
@@ -1,0 +1,19 @@
+package io.micronaut.test.lombok;
+
+import io.micronaut.core.beans.BeanIntrospection;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class LombokIntrospectedBuilderTest {
+
+    @Test
+    void testLombokBuilder() {
+        BeanIntrospection.Builder<RobotEntity> builder = BeanIntrospection.getIntrospection(RobotEntity.class)
+            .builder();
+
+        RobotEntity robotEntity = builder.with("name", "foo")
+            .build();
+
+        Assertions.assertEquals("foo", robotEntity.getName());
+    }
+}

--- a/test-suite/src/test/java/io/micronaut/test/lombok/RobotEntity.java
+++ b/test-suite/src/test/java/io/micronaut/test/lombok/RobotEntity.java
@@ -1,0 +1,16 @@
+package io.micronaut.test.lombok;
+
+import io.micronaut.core.annotation.Introspected;
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder(builderClassName = "Builder")
+@Introspected(builder =
+    @Introspected.IntrospectionBuilder(
+        builderClass = RobotEntity.Builder.class
+    )
+)
+public class RobotEntity {
+    String name;
+}

